src/test_utils.jl

```diff
@@ -160,13 +160,26 @@ using Mooncake:
     NoRData,
     rdata_type,
     rdata
+using Preferences: load_preference, get_uuid
+using DispatchDoctor: type_instability, allow_unstable as _allow_unstable
 
 struct Shim end
 
-test_opt(x...) = test_opt_internal(Shim(), x...)
+const DD_ENABLED = let uuid = get_uuid(@__MODULE__)
+    mode = load_preference(uuid, "dispatch_doctor_mode")
+
+    mode ∉ (nothing, "disable")
+end
```

Comment:
I am concerned about 

1. DD might quietly switch off the JET and AllocChecks tests. 
2. The changes make this test-related code hard to read.  

Instead, for clarity, can we introduce 

```julia
struct JETTests end
struct DispatchDoctorTests end
struct AllocChecksTests end

test_opt(test_ctx::JETTests, x...) = test_opt_internal(Shim(), x...)
test_opt(test_ctx::DispatchDoctorTests, x...) =  nothing
check_allocs(test_ctx::DispatchDoctorTests, ...) = nothing
```

Then, users must explicitly pass the `test_ctx` context to calls to `test_opt` and `check_allocs`.

Reply:
I like that design, sure.

Reply:
@yebai I just tried this. It makes the TestUtils implementation extremely complex, as you have to pass the context through many layers of functions, and track it. Keep in mind that the DispatchDoctor tests are running the same test code as usual.

I think this refactor would only be worth it if there were multiple additional interacting contexts.

Reply:
> Should they be made to include(joinpath(@__DIR__, "..", "utils.jl")) and so forth? The tricky thing is that they would still need all the same imports as for basic.

It's okay to include these files from integration tests. The key part is a separate environment.

Reply:
Makes sense!

Ok i have made the change.